[move-cli] Print resource changes after verbose execution

darioncassel opened the pull request:
## Motivation

This is in response to issue #8515. The goal is to print resource changes on execution Move sandbox commands with the verbose flag active. This removes a step where resource changes must be manually inspected with the `sandbox view` command. 

See below of example output for the three possible resource changes (`+` indicates diff, not part of actual output).

Addition now shows the resource that was added:
```diff
 Changed resource(s) under 1 address(es):
   Changed 1 resource(s) under address 0000000000000000000000000000000F:
     Added type 0x2::Test::Resource: [10, 0, 0, 0, 0, 0, 0, 0] (wrote 40 bytes)
+      key 0x2::Test::Resource {
+           i: 10
+      }
```

Modification now shows the previous and new states of the resource:
```diff
 Changed resource(s) under 1 address(es):
   Changed 1 resource(s) under address 0000000000000000000000000000000F:
     Changed type 0x2::Test::Resource: [20, 0, 0, 0, 0, 0, 0, 0] (wrote 40 bytes)
+      Previous:
+        key 0x2::Test::Resource {
+            i: 10
+        }
+      New:
+        key 0x2::Test::Resource {
+            i: 20
+        }
```

Deletion now shows the resource that was removed:
```diff
 Changed resource(s) under 1 address(es):
   Changed 1 resource(s) under address 0000000000000000000000000000000F:
     Deleted type 0x2::Test::Resource (wrote 32 bytes)
+      key 0x2::Test::Resource {
+          i: 10
+      }
```

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

This is covered via `cargo test`. To account for the output changes, the `diem_smoke` and `events_emit_view` tests' expected output have been updated.

## Related PRs

N/A

## If targeting a release branch, please fill the below out as well

N/A

mengxu-fb commented on language/tools/move-cli/src/sandbox/utils/mod.rs:
```diff
@@ -82,6 +83,19 @@ pub(crate) fn explain_publish_changeset(changeset: &ChangeSet, state: &OnDiskSta
+    let mut indent_str: String = "".to_owned();
+    for _i in 0..indent {
+        indent_str.push(' ');
+    }
```
I think this might be simplified to `(0..indent).map(|_| " ").collect::<String>();`

darioncassel replied:
Good idea; I'll fix that.

mengxu-fb reviewed (approved):
LGTM! Double checking with @metajack to see whether this PR addresses the issue you raised on Slack the other day.

metajack commented:
This looks amazing!

When you show the previous and new versions on a change, do you show the whole object or just the parts that changed? Obviously awesome either way, but if it's the whoel object, it might be a nice future extension to just show the changed parts and elide the others.

darioncassel commented:
@metajack Thanks! Currently the entire previous and new object are printed. It's definitely possible to show the changes in a more compact form.

I came up with prototypes for three options. Here's what they look like using the `diem_smoke` test.

1) We print every unchanged line, but show a diff for the changed line:
```
  Changed 1 resource(s) under address 0000000000000000000000000A550C18:
    Changed type 0x1::DiemAccount::AccountOperationsCapability: [0, 4, 0, 0, 0, 0, 0, 0, 0, 24, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 10, 85, 12, 24] (wrote 92 bytes)
        key 0x1::DiemAccount::AccountOperationsCapability {
            limits_cap: store 0x1::AccountLimits::AccountLimitMutationCapability {
                dummy_field: false
            }
            creation_events: store 0x1::Event::EventHandle<0x1::DiemAccount::CreateAccountEvent> {
-                counter: 3
+                counter: 4
                guid: 00000000000000000000000000000000000000000a550c18
            }
        }
```

2) As you were suggesting, we only print the changes (and the top-level name of the resource):
```
  Changed 1 resource(s) under address 0000000000000000000000000A550C18:
    Changed type 0x1::DiemAccount::AccountOperationsCapability: [0, 4, 0, 0, 0, 0, 0, 0, 0, 24, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 10, 85, 12, 24] (wrote 92 bytes)
        key 0x1::DiemAccount::AccountOperationsCapability {
        ...
-                counter: 3
+                counter: 4
        ...
        }
```

3) Similar to the second option but we only print the new value:
```
  Changed 1 resource(s) under address 0000000000000000000000000A550C18:
    Changed type 0x1::DiemAccount::AccountOperationsCapability: [0, 4, 0, 0, 0, 0, 0, 0, 0, 24, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 10, 85, 12, 24] (wrote 92 bytes)
        key 0x1::DiemAccount::AccountOperationsCapability {
        ...
                counter: 4
        ...
        }
```

All of these options are more compact than what we had before, but I worry that we might lose a little too much context with the second and third options? Let me know what you think.

mengxu-fb commented:
I think the first option aligns better with what we see on the exp files, i.e., when a test output differs from the exp file, a diff is printed.

And thinking on this line, we might be able to represent resource addition, deletion, and update with `+`, `-`, `=` prefixes:
e.g.,

Resource addition
```diff
+ key 0x2::M::R {
+   value: 0;
+   flag: false;
+ }
```

Resource deletion
```diff
- key 0x2::M::R {
-   value: 0;
-   flag: false;
- }
```

Resource update
```diff
= key 0x2::M::R {
-   value: 0;
+   value: 1;
=   flag: false;
= }
```

sblackshear commented:
+1 on the awesomeness of this!

If we decide to go with diff printing in the future, the `difference` crate used in the compiler may be helpful (assuming you're not already using it!) https://github.com/diem/diem/blob/main/language/move-lang/tests/move_check_testsuite.rs#L22

metajack commented:
I prefer 2, although I think it should print the intermediate struct there omitted other fields. That way you see the whole path to the change.

If you go with 1, large structs will generate a lot of noise, and either is sufficient for exp files to check output hasn't changed I believe.

I don't think I have a strong opinion on showing the old value vs the new one. I suppose it looks nice when it's a primitive field, but if you replace a whole struct, it might also get really noisy. It may be something we need to revisit later once we have more usage experience.

darioncassel commented:
@sblackshear Thanks for the feedback! Unfortunately I had already written a simple diff by the time I saw your comment 😆 
But I can definitely switch it over to use the `difference` crate for maintainability.

@metajack The problem with printing the full path to the changed resource is that I'm doing a textual diff rather than a structural one, and so it's difficult to see the differences semantically. The reason for this is that I'm piggybacking off of the existing support for pretty printing Move structs. It's definitely possible to do a structural diff and print from the results of that but I worry that it will require modification or reimplementation of what the `resource_viewer` crate is doing, which might be out of the scope of this PR. Let me know if you know of a simpler way.

darioncassel commented:
After discussion with @mengxu-fb we've decided to leave diffs of changes to a future PR. The reason is that the most suitable form for those diffs is not known at this time and will likely require further discussion with the team to get right.

/land
